### PR TITLE
feat: add new error types ( active_campaign, adjust, adobe_analytics, appsflyer, airship)

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

Destination List

- Active Campaign
- Adjust
- Adobe Analytics
- Apps Flyer
- Airship

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/active_campaign/util.js [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
-const { CustomError } = require("../../util");
+const {
+  RetryableError,
+  NetworkInstrumentationError
+} = require("../../util/errorTypes");
 
 const errorHandler = (err, message) => {
   if (err.response) {
-    throw new CustomError(
+    throw new NetworkInstrumentationError(
```
Comment (saikumarrs) [inline src/v0/destinations/active_campaign/util.js:11]: How do we know instrumentation is the problem?

File: src/v0/destinations/active_campaign/util.js [hunk 2]
```diff
@@ -1,19 +1,24 @@
...
-      err.response.status || 400
+      )})`
     );
   } else {
     const httpError = nodeSysErrorToStatus(err.code);
-    throw new CustomError(`${message} ${httpError.message}`, httpError.status);
+    if (httpError.status === 500) {
+      throw new RetryableError(`${message} ${httpError.message}`);
+    }
+    throw new NetworkInstrumentationError(`${message} ${httpError.message}`);
```
Comment (saikumarrs) [inline src/v0/destinations/active_campaign/util.js:21]: Is this accurate w.r.t the docs?

**Comment by ujjwal-ab** [comment]
> Is test data not impact with these changes?

No, As the error message is same for all.

File: src/v0/destinations/active_campaign/util.js (hunk 2 above)
Comment (ujjwal-ab) [reply src/v0/destinations/active_campaign/util.js:21]: In `networkUtils`, we have a mapping of `ReceivedCode-> status`, for 500 status we can do retryable. We are returning 400 [here](https://github.com/rudderlabs/rudder-transformer/blob/23ce982f2756c62d009250aba3cafc87f4b63b67/src/adapters/utils/networkUtils.js#L15) in multiple scenarios.

File: src/v0/destinations/active_campaign/transform.js [hunk 3]
```diff
@@ -435,7 +440,7 @@ const screenRequestHandler = async (message, category, destination) => {
   }
 
   if (res.response.status !== 200)
-    throw new CustomError("Unable to create event", res.response.status || 400);
+    throw new UnauthorizedError("Unable to create event");
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/active_campaign/transform.js:443]: Actually not, here it is coming if `response.success != false`, probably `NetworkInstrumentation` will be the best suit.

File: src/v0/destinations/active_campaign/util.js (hunk 1 above)
Comment (ujjwal-ab) [reply src/v0/destinations/active_campaign/util.js:11]: Here, if we are getting a false success Response for an axios call. This only seems the best suite.

Comment (saikumarrs) [reply src/v0/destinations/active_campaign/util.js:11]: Instrumentation error puts blame on the user. Unless we know for sure, we shouldn't do that.
You can use `AbortedError` and pass the status code to it.

Comment (saikumarrs) [reply src/v0/destinations/active_campaign/util.js:11]: Actually, use `NetworkError` and deduce the error type via `getDynamicErrorType` method. See the parent PR for reference.

File: src/v0/destinations/active_campaign/transform.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (saikumarrs) [reply src/v0/destinations/active_campaign/transform.js:443]: Same as the other comment.

[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/active_campaign/transform.js]: Updated

File: src/v0/destinations/active_campaign/transform.js [hunk 4]
```diff
@@ -71,7 +72,14 @@ const syncContact = async (contactPayload, category, destination) => {
   }
   const createdContact = get(res, "response.data.contact"); // null safe
   if (!createdContact) {
-    throw new NetworkInstrumentationError("Unable to Create Contact");
+    throw new NetworkError(
+      "Unable to Create Contact",
+      res.response.status,
```
Comment (saikumarrs) [inline src/v0/destinations/active_campaign/transform.js:77]:
```suggestion
      res.response?.status,
```

File: src/v0/destinations/active_campaign/transform.js [hunk 5]
```diff
@@ -71,7 +72,14 @@ const syncContact = async (contactPayload, category, destination) => {
...
+      {
+        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response.status)
+      },
+      res?.response?.data
```
Comment (saikumarrs) [inline src/v0/destinations/active_campaign/transform.js:81]:
```suggestion
      res.response?.data
```

File: src/v0/destinations/active_campaign/transform.js [hunk 6]
```diff
@@ -71,7 +72,14 @@ const syncContact = async (contactPayload, category, destination) => {
   }
   const createdContact = get(res, "response.data.contact"); // null safe
   if (!createdContact) {
-    throw new NetworkInstrumentationError("Unable to Create Contact");
+    throw new NetworkError(
+      "Unable to Create Contact",
+      res.response.status,
+      {
+        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response.status)
```
Comment (saikumarrs) [inline src/v0/destinations/active_campaign/transform.js:79]:
```suggestion
        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response?.status)
```

File: src/v0/destinations/active_campaign/transform.js [hunk 7]
```diff
@@ -439,8 +447,16 @@ const screenRequestHandler = async (message, category, destination) => {
...
+  if (res.response.status !== 200) {
+    throw new NetworkError(
+      "Unable to create event",
+      res.response.status,
+      {
+        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response.status)
+      },
+      res?.response?.data
+    );
+  }
```
Comment (saikumarrs) [inline src/v0/destinations/active_campaign/transform.js:459]: Use optional chaining

File: src/v0/destinations/af/transform.js [hunk 8]
```diff
@@ -1,4 +1,5 @@
 const get = require("get-value");
+const { padStart } = require("lodash");
```
Comment (saikumarrs) [inline src/v0/destinations/af/transform.js:2]: ?

File: src/v0/destinations/active_campaign/transform.js [hunk 9]
```diff
@@ -529,9 +559,13 @@ const trackRequestHandler = async (message, category, destination) => {
     };
     res = await httpPOST(endpoint, requestData, requestOpt);
     if (res.response.status !== 201) {
```
Comment (koladilip) [inline src/v0/destinations/active_campaign/transform.js:561]: res?.response?.status

File: src/v0/destinations/airship/transform.js [hunk 10]
```diff
@@ -203,9 +201,8 @@ const groupResponseBuilder = (message, { Config }) => {
 
   const traits = flattenJson(getFieldValueFromMessage(message, "traits"));
   if (!isDefinedAndNotNullAndNotEmpty(traits)) {
-    throw new CustomError(
-      "[Airship]:: for group, tags or attributes properties are required under traits",
-      400
+    throw new InstrumentationError(
+      "[Airship]:: for group, tags or attributes properties are required under traits"
```
Comment (koladilip) [inline src/v0/destinations/airship/transform.js:205]: remove [Airship]:: prefix for everywhere

File: src/v0/destinations/active_campaign/transform.js [hunk 11]
```diff
@@ -66,7 +72,14 @@ const syncContact = async (contactPayload, category, destination) => {
...
   const createdContact = get(res, "response.data.contact"); // null safe
   if (!createdContact) {
-    throw new CustomError("Unable to Create Contact", 400);
+    throw new NetworkError(
+      "Unable to Create Contact",
+      res.response?.status,
+      {
+        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response?.status)
+      },
+      res.response?.data
```
Comment (utsabc) [inline src/v0/destinations/active_campaign/transform.js:81]:
```suggestion
      res.response
```

File: src/v0/destinations/active_campaign/transform.js [hunk 12]
```diff
@@ -434,8 +447,16 @@ const screenRequestHandler = async (message, category, destination) => {
...
-  if (res.response.status !== 200)
-    throw new CustomError("Unable to create event", res.response.status || 400);
+  if (res?.response?.status !== 200) {
+    throw new NetworkError(
+      "Unable to create event",
+      res.response?.status,
+      {
+        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response?.status)
+      },
+      res.response?.data
```
Comment (utsabc) [inline src/v0/destinations/active_campaign/transform.js:457]:
```suggestion
      res.response
```

File: src/v0/destinations/active_campaign/transform.js [hunk 13]
```diff
@@ -464,9 +485,13 @@ const screenRequestHandler = async (message, category, destination) => {
...
     if (res.response.status !== 201) {
-      throw new CustomError(
+      throw new NetworkError(
         "Unable to create event",
-        res.response.status || 400
+        res.response.status,
+        {
+          [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response.status)
+        },
+        res?.response?.data
```
Comment (utsabc) [inline src/v0/destinations/active_campaign/transform.js:494]:
```suggestion
        res?.response
```

File: src/v0/destinations/active_campaign/transform.js [hunk 14]
```diff
@@ -500,11 +525,16 @@ const trackRequestHandler = async (message, category, destination) => {
...
-    throw new CustomError(
+  if (res.response.status !== 200) {
+    throw new NetworkError(
       "Unable to fetch events. Aborting",
-      res.response.status || 400
+      res.response.status,
+      {
+        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response.status)
+      },
+      res?.response?.data
```
Comment (utsabc) [inline src/v0/destinations/active_campaign/transform.js:535]:
```suggestion
      res?.response
```

File: src/v0/destinations/active_campaign/transform.js [hunk 15]
```diff
@@ -528,10 +558,14 @@ const trackRequestHandler = async (message, category, destination) => {
...
-      throw new CustomError(
+    if (res?.response?.status !== 201) {
+      throw new NetworkError(
         "Unable to create event. Aborting",
-        res.response.status || 400
+        res.response.status,
+        {
+          [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(res.response.status)
+        },
+        res?.response?.data
```
Comment (utsabc) [inline src/v0/destinations/active_campaign/transform.js:568]:
```suggestion
        res?.response
```

**Review by utsabc — APPROVED** [review]

File: src/v0/destinations/af/deleteUsers.js [hunk 16]
```diff
@@ -74,21 +72,15 @@ const userDeletionHandler = async (userAttributes, config) => {
         userAttributes[i].appsflyer_id
       );
       if (!response || !response.response) {
-        throw new ErrorBuilder()
-          .setMessage("Could not get response")
-          .setStatus(500)
-          .build();
+        throw new TransformationError("Could not get response");
```
Comment (saikumarrs) [inline src/v0/destinations/af/deleteUsers.js:75]: Not a transformation error.

**Review by saikumarrs — CHANGES_REQUESTED** [review]
TransformationError is incorrectly used.

File: src/v0/destinations/af/deleteUsers.js (hunk 16 above)
Comment (ujjwal-ab) [reply src/v0/destinations/af/deleteUsers.js:75]: Should it be `RetryableError`?

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**